### PR TITLE
Add water management and inundation models to MOSART

### PR DESCRIPTION
**Pull request by hydrotian**
Add water management and inundation models to MOSART.

[BFB]

### CODE REVIEW

**Comment by rljacob** [comment]
Confluence page:  https://acme-climate.atlassian.net/wiki/spaces/ED/pages/981074560/W3+Water+management+and+flood+inundation

File: components/mosart/bld/namelist_files/namelist_defaults_mosart.xml [hunk 1]
```diff
@@ -32,4 +39,34 @@ for the CLM data in the CESM distribution
 <frivinp_rtm rof_grid="r05" >rof/mosart/MOSART_global_half_20180721a.nc</frivinp_rtm>
 <frivinp_rtm rof_grid="NLDAS" >rof/mosart/MOSART_NLDAS_8th.nc</frivinp_rtm>
 
+<!-- WRM -->
+
+<paraFile rof_grid="NLDAS" >rof/mosart/US_reservoir_8th_NLDAS3_c20161220_updated_20170314.nc</paraFile>
```
Comment (bishtgautam) [inline components/mosart/bld/namelist_files/namelist_defaults_mosart.xml]: this file needs to be uploaded to E3SM input data server

File: components/mosart/bld/namelist_files/namelist_defaults_mosart.xml [hunk 2]
```diff
@@ -32,4 +39,34 @@ for the CLM data in the CESM distribution
...
+<demandPath rof_grid="NLDAS" >/pic/scratch/vois549/wm_data/NLDAS2_GCAM_water_demand_</demandPath>
```
Comment (bishtgautam) [inline components/mosart/bld/namelist_files/namelist_defaults_mosart.xml]: Why is this path hard-coded for PNNL machine? Also, the path should be relative to the inpudata dir for various machines.

File: components/mosart/src/cpl/rof_comp_esmf.F90 [hunk 3]
```diff
@@ -691,11 +695,24 @@ subroutine rof_import_esmf( x2r_array, rc)
        else
           rtmCTL%qdto(n,nliq) = 0.0_r8
        endif
+ !      rtmCTL%qdem(n,nliq) = fptr(index_x2r_Flrl_demand,n2) * (rtmCTL%area(n)*0.001_r8)
 
        rtmCTL%qsur(n,nfrz) = fptr(index_x2r_Flrl_rofi,n2) * (rtmCTL%area(n)*0.001_r8)
        rtmCTL%qsub(n,nfrz) = 0.0_r8
        rtmCTL%qgwl(n,nfrz) = 0.0_r8
        rtmCTL%qdto(n,nfrz) = 0.0_r8
+       !rtmCTL%qdem(n,nfrz) = 0.0_r8
```
Comment (bishtgautam) [inline components/mosart/src/cpl/rof_comp_esmf.F90]: Please delete the commented out lines

File: components/mosart/src/cpl/rof_comp_esmf.F90 [hunk 4]
```diff
@@ -691,11 +695,24 @@ subroutine rof_import_esmf( x2r_array, rc)
        else
           rtmCTL%qdto(n,nliq) = 0.0_r8
        endif
+ !      rtmCTL%qdem(n,nliq) = fptr(index_x2r_Flrl_demand,n2) * (rtmCTL%area(n)*0.001_r8)
```
Comment (bishtgautam) [inline components/mosart/src/cpl/rof_comp_esmf.F90]: Please delete the commented out line

File: components/mosart/src/cpl/rof_comp_esmf.F90 [hunk 5]
```diff
@@ -445,7 +449,7 @@ subroutine rof_run_esmf(comp, import_state, export_state, EClock, rc)
     write(rdate,'(i4.4,"-",i2.2,"-",i2.2,"-",i5.5)') yr_sync,mon_sync,day_sync,tod_sync
     nlend = seq_timemgr_StopAlarmIsOn( EClock )
     rstwr = seq_timemgr_RestartAlarmIsOn( EClock )
-    call advance_timestep()
+    !call advance_timestep()
```
Comment (bishtgautam) [inline components/mosart/src/cpl/rof_comp_esmf.F90]: Please delete the commented out line

File: components/mosart/src/cpl/rof_comp_esmf.F90 [hunk 6]
```diff
@@ -797,6 +814,14 @@ subroutine rof_export_esmf( r2x_array, rc)
        fptr(index_r2x_Flrr_flood,ni) = -rtmCTL%flood(n)/(rtmCTL%area(n)*0.001_r8)
        fptr(index_r2x_Flrr_volr,ni)    = (Trunoff%wr(n,nliq) + Trunoff%wt(n,nliq)) / rtmCTL%area(n)
        fptr(index_r2x_Flrr_volrmch,ni) = Trunoff%wr(n,nliq) / rtmCTL%area(n)
+ !      fptr(index_r2x_Flrr_supply,ni)  = 0._r8  ! tcxcpl
```
Comment (bishtgautam) [inline components/mosart/src/cpl/rof_comp_esmf.F90]: Please delete the commented out lines

File: components/mosart/src/cpl/rof_comp_mct.F90 [hunk 7]
```diff
@@ -343,7 +347,7 @@ subroutine rof_run_mct( EClock, cdata_r, x2r_r, r2x_r)
     write(rdate,'(i4.4,"-",i2.2,"-",i2.2,"-",i5.5)') yr_sync,mon_sync,day_sync,tod_sync
     nlend = seq_timemgr_StopAlarmIsOn( EClock )
     rstwr = seq_timemgr_RestartAlarmIsOn( EClock )
-    call advance_timestep()
+    !call advance_timestep()
```
Comment (bishtgautam) [inline components/mosart/src/cpl/rof_comp_mct.F90]: Please delete the commented out line

File: components/mosart/src/cpl/rof_comp_mct.F90 [hunk 8]
```diff
@@ -591,11 +595,25 @@ subroutine rof_import_mct( x2r_r)
        else
           rtmCTL%qdto(n,nliq) = 0.0_r8
        endif
+!       rtmCTL%qdem(n,nliq) = x2r_r%rAttr(index_x2r_Flrl_demand,n2) * (rtmCTL%area(n)*0.001_r8)
```
Comment (bishtgautam) [inline components/mosart/src/cpl/rof_comp_mct.F90]: Please delete the commented out line

File: components/mosart/src/cpl/rof_comp_mct.F90 [hunk 9]
```diff
@@ -591,11 +595,25 @@ subroutine rof_import_mct( x2r_r)
...
 
        rtmCTL%qsur(n,nfrz) = x2r_r%rAttr(index_x2r_Flrl_rofi,n2) * (rtmCTL%area(n)*0.001_r8)
        rtmCTL%qsub(n,nfrz) = 0.0_r8
        rtmCTL%qgwl(n,nfrz) = 0.0_r8
        rtmCTL%qdto(n,nfrz) = 0.0_r8
+!       rtmCTL%qdem(n,nfrz) = 0.0_r8
```
Comment (bishtgautam) [inline components/mosart/src/cpl/rof_comp_mct.F90]: Please delete the commented out lines

File: components/mosart/src/cpl/rof_comp_mct.F90 [hunk 10]
```diff
@@ -692,6 +710,14 @@ subroutine rof_export_mct( r2x_r )
        r2x_r%rattr(index_r2x_Flrr_flood,ni)   = -rtmCTL%flood(n) / (rtmCTL%area(n)*0.001_r8)
        r2x_r%rattr(index_r2x_Flrr_volr,ni)    = (Trunoff%wr(n,nliq) + Trunoff%wt(n,nliq)) / rtmCTL%area(n)
        r2x_r%rattr(index_r2x_Flrr_volrmch,ni) = Trunoff%wr(n,nliq) / rtmCTL%area(n)
+!       r2x_r%rattr(index_r2x_Flrr_supply,ni)  = 0._r8  ! tcxcpl
```
Comment (bishtgautam) [inline components/mosart/src/cpl/rof_comp_mct.F90]: Please delete the commented out lines

File: components/mosart/src/cpl/rof_cpl_indices.F90 [hunk 11]
```diff
@@ -28,12 +28,25 @@ module rof_cpl_indices
   integer, public :: index_x2r_Flrl_rofsub = 0  ! lnd->rof liquid subsurface runoff from land
   integer, public :: index_x2r_Flrl_rofdto = 0  ! lnd->rof liquid direct to ocean runoff
   integer, public :: index_x2r_Flrl_rofi  = 0   ! lnd->rof ice runoff forcing from land
+!  integer, public :: index_x2r_Flrl_demand = 0  ! lnd->rof input total fluxes (<= 0)
```
Comment (bishtgautam) [inline components/mosart/src/cpl/rof_cpl_indices.F90]: Please delete the commented out lines

File: components/mosart/src/cpl/rof_cpl_indices.F90 [hunk 12]
```diff
@@ -42,6 +55,8 @@ module rof_cpl_indices
   integer, public :: index_r2x_Flrr_flood = 0   ! rof->lnd flood runoff (>fthresh) back to land
   integer, public :: index_r2x_Flrr_volr = 0    ! rof->lnd volr total volume back to land
   integer, public :: index_r2x_Flrr_volrmch = 0 ! rof->lnd volr main channel back to land
+  !integer, public :: index_r2x_Flrr_supply = 0  ! rof->lnd supply flux for land use
```
Comment (bishtgautam) [inline components/mosart/src/cpl/rof_cpl_indices.F90]: Please delete the commented out lines

File: components/mosart/src/cpl/rof_cpl_indices.F90 [hunk 13]
```diff
@@ -83,6 +98,17 @@ subroutine rof_cpl_indices_set( )
     index_x2r_Flrl_rofsub = mct_avect_indexra(avtmp,'Flrl_rofsub')
     index_x2r_Flrl_rofdto = mct_avect_indexra(avtmp,'Flrl_rofdto',perrwith='quiet')
     index_x2r_Flrl_rofi   = mct_avect_indexra(avtmp,'Flrl_rofi')
+!    index_x2r_Flrl_demand = mct_avect_indexra(avtmp,'Flrl_demand')
```
Comment (bishtgautam) [inline components/mosart/src/cpl/rof_cpl_indices.F90]: Please delete the commented out lines

File: components/mosart/src/cpl/rof_cpl_indices.F90 [hunk 14]
```diff
@@ -97,7 +123,8 @@ subroutine rof_cpl_indices_set( )
     index_r2x_Flrr_flood = mct_avect_indexra(avtmp,'Flrr_flood')
     index_r2x_Flrr_volr  = mct_avect_indexra(avtmp,'Flrr_volr')
     index_r2x_Flrr_volrmch = mct_avect_indexra(avtmp,'Flrr_volrmch')
-
+    !index_r2x_Flrr_supply = mct_avect_indexra(avtmp,'Flrr_supply')
```
Comment (bishtgautam) [inline components/mosart/src/cpl/rof_cpl_indices.F90]: Please delete the commented out lines

File: components/mosart/src/riverroute/RtmHistFlds.F90 [hunk 15]
```diff
@@ -126,6 +146,61 @@ subroutine RtmHistFldsInit()
...
+         avgflag='A', long_name='MOSART total demand: '//trim(rtm_tracers(2)), &
+         ptr_rof=rtmCTL%qdem_nt2, default='active')
+
+!#ifdef INCLUDE_WRM
+    if (wrmflag) then
+
+      call RtmHistAddfld (fname='WRM_SUPPLY', units='m3/s',  &
+         avgflag='A', long_name='WRM supply provided ', &
+         ptr_rof=StorWater%Supply, default='active')
+      !call RtmHistAddfld (fname='WRM_SUPPLY_FRACTION', units='none',  &
```
Comment (bishtgautam) [inline components/mosart/src/riverroute/RtmHistFlds.F90]: Please delete the commented out lines

File: components/mosart/src/riverroute/RtmRestFile.F90 [hunk 16]
```diff
@@ -376,14 +380,36 @@ subroutine RtmRestart(ncid, flag)
...
     character(len=32)  :: vname,uname
     character(len=128) :: lname
     !-----------------------------------------------------------------------
 
-    do nv = 1,7
+    compute_release = .false.
+    storage_read = .true.
+    release_read = .true.
+    stormth_read = .true.
+!#ifdef INCLUDE_WRM
```
Comment (bishtgautam) [inline components/mosart/src/riverroute/RtmRestFile.F90]: Please delete the commented out lines

File: components/mosart/src/riverroute/RtmRestFile.F90 [hunk 17]
```diff
@@ -419,11 +445,61 @@ subroutine RtmRestart(ncid, flag)
           lname = 'instataneous flow out of main channel in cell'
           uname = 'm3/s'
           dfld  => rtmCTL%erout(:,nt)
+!#ifdef INCLUDE_WRM
```
Comment (bishtgautam) [inline components/mosart/src/riverroute/RtmRestFile.F90]: Please delete the commented out lines

File: components/mosart/src/riverroute/RtmRestFile.F90 [hunk 18]
```diff
@@ -453,7 +539,14 @@ subroutine RtmRestart(ncid, flag)
              if (abs(rtmCTL%wt(n,nt))      > 1.e30) rtmCTL%wt(n,nt) = 0.
              if (abs(rtmCTL%wr(n,nt))      > 1.e30) rtmCTL%wr(n,nt) = 0.
              if (abs(rtmCTL%erout(n,nt))   > 1.e30) rtmCTL%erout(n,nt) = 0.
-          end do
+          end do  ! nt
+!#ifdef INCLUDE_WRM
```
Comment (bishtgautam) [inline components/mosart/src/riverroute/RtmRestFile.F90]: Please delete the commented out lines

File: components/mosart/src/riverroute/RtmRestFile.F90 [hunk 19]
```diff
@@ -465,8 +558,23 @@ subroutine RtmRestart(ncid, flag)
                 rtmCTL%dvolrdtocn(n,nt)= rtmCTL%dvolrdt(n,nt)
              enddo
           endif
-       enddo
-    endif
+       enddo  ! n
+
+!#ifdef INCLUDE_WRM
```
Comment (bishtgautam) [inline components/mosart/src/riverroute/RtmRestFile.F90]: Please delete the commented out lines

File: components/mosart/src/cpl/rof_comp_esmf.F90 [hunk 20]
```diff
@@ -31,14 +31,18 @@ module rof_comp_esmf
                                 inst_index, inst_suffix, inst_name, RtmVarSet
   use RtmSpmd          , only : masterproc, iam, npes, RtmSpmdInit, ROFID
   use RtmMod           , only : Rtmini, Rtmrun
-  use RtmTimeManager   , only : timemgr_setup, get_curr_date, get_step_size, advance_timestep 
+  use RtmTimeManager   , only : timemgr_setup, get_curr_date, get_step_size!, advance_timestep 
+!#ifdef INCLUDE_WRM
```
Comment (bishtgautam) [inline components/mosart/src/cpl/rof_comp_esmf.F90]: Please delete the commented out lines

File: components/mosart/src/riverroute/MOSART_physics_mod.F90 [hunk 21]
```diff
@@ -46,9 +54,58 @@ subroutine Euler
...
+    integer :: iunit, idam, m, k, unitUp, cnt, ier   !local index
     real(r8) :: temp_erout, localDeltaT
     real(r8) :: negchan
+    integer  :: yr,mon,day,tod
+    character(len=*),parameter :: subname = '(Euler)'
+    !------------------
+
+    call get_curr_date(yr, mon, day, tod)
+	
+!#if (1 == 0)
```
Comment (bishtgautam) [inline components/mosart/src/riverroute/MOSART_physics_mod.F90]: Please delete the commented out lines

File: components/mosart/src/riverroute/MOSART_physics_mod.F90 [hunk 22]
```diff
@@ -46,9 +54,58 @@ subroutine Euler
...
+!  if (masterproc) write(iulog,*) trim(subname), 'Tian testing hashtag comment working'
+!#endif
+!  if (masterproc) write(iulog,*) trim(subname), 'Tian testing hashtag comment not working'
+!  write(iulog,*) trim(subname),' updating monthly data at ',yr,mon,day,tod
+    !------------------
+    ! WRM prep
+    !------------------
+
+!#ifdef INCLUDE_WRM
```
Comment (bishtgautam) [inline components/mosart/src/riverroute/MOSART_physics_mod.F90]: Please delete the commented out lines

File: components/mosart/src/riverroute/MOSART_physics_mod.F90 [hunk 23]
```diff
@@ -62,29 +119,39 @@ subroutine Euler
...
+! NV WARNNG WARNING JUST FOR TESTING
+!          TRunoff%qsub(iunit,nt) = TRunoff%qsub(iunit,nt) / 1.6_r8
           TRunoff%etin(iunit,nt) = (-TRunoff%ehout(iunit,nt) + TRunoff%qsub(iunit,nt)) * TUnit%area(iunit) * TUnit%frac(iunit)
        endif
     end do
     endif
     end do
     call t_stopf('mosartr_hillslope')
 
+!moved inside the subnetwork channel routing and work on wt instead of etin
```
Comment (bishtgautam) [inline components/mosart/src/riverroute/MOSART_physics_mod.F90]: Please delete the commented out lines

File: components/mosart/src/riverroute/MOSART_physics_mod.F90 [hunk 24]
```diff
@@ -62,29 +119,39 @@ subroutine Euler
           call hillslopeRouting(iunit,nt,Tctl%DeltaT)
           TRunoff%wh(iunit,nt) = TRunoff%wh(iunit,nt) + TRunoff%dwh(iunit,nt) * Tctl%DeltaT
           call UpdateState_hillslope(iunit,nt)
+! NV WARNNG WARNING JUST FOR TESTING
```
Comment (bishtgautam) [inline components/mosart/src/riverroute/MOSART_physics_mod.F90]: Please delete the commented out lines

File: components/mosart/src/riverroute/MOSART_physics_mod.F90 [hunk 25]
```diff
@@ -182,6 +264,39 @@ subroutine Euler
...
+                         call insert_returnflow_channel(iunit, localDeltaT )
+                      endif
+                      ! update main channel storage as well
+                      temp_erout = temp_erout - TRunoff%erout(iunit,nt) ! change in erout after regulation and extraction
+                      TRunoff%dwr(iunit,nt) =  temp_erout
+                      TRunoff%wr(iunit,nt) = TRunoff%wr(iunit,nt) + TRunoff%dwr(iunit,nt) * localDeltaT
+                      call UpdateState_mainchannel(iunit,nt)
+                   endif
+! tcraig, moved out of loop
+                   !if ( ctlSubwWRM%RegulationFlag>0 .and. WRMUnit%INVicell(iunit) > 0 .and. WRMUnit%MeanMthFlow(iunit,13) > 0.01_r8 ) then
```
Comment (bishtgautam) [inline components/mosart/src/riverroute/MOSART_physics_mod.F90]: Please delete the commented out lines

File: components/mosart/src/riverroute/MOSART_physics_mod.F90 [hunk 26]
```diff
@@ -182,6 +264,39 @@ subroutine Euler
...
+                      if (ctlSubwWRM%TotalDemandFlag > 0 .AND. ctlSubwWRM%ReturnFlowFlag > 0 ) then
+                         call insert_returnflow_channel(iunit, localDeltaT )
+                      endif
+                      ! update main channel storage as well
+                      temp_erout = temp_erout - TRunoff%erout(iunit,nt) ! change in erout after regulation and extraction
+                      TRunoff%dwr(iunit,nt) =  temp_erout
+                      TRunoff%wr(iunit,nt) = TRunoff%wr(iunit,nt) + TRunoff%dwr(iunit,nt) * localDeltaT
+                      call UpdateState_mainchannel(iunit,nt)
+                   endif
+! tcraig, moved out of loop
```
Comment (bishtgautam) [inline components/mosart/src/riverroute/MOSART_physics_mod.F90]: Please delete the commented out lines

File: components/mosart/src/riverroute/MOSART_physics_mod.F90 [hunk 27]
```diff
@@ -686,23 +881,25 @@ subroutine printTest(nio)
...
                      TRunoff%etin(IDlist(3),1)/TUnit%area(IDlist(3)), TRunoff%erlateral(IDlist(3),1)/TUnit%area(IDlist(3)), TRunoff%flow(IDlist(3),1), &
                      TRunoff%etin(IDlist(4),1)/TUnit%area(IDlist(4)), TRunoff%erlateral(IDlist(4),1)/TUnit%area(IDlist(4)), TRunoff%flow(IDlist(4),1), &
                      TRunoff%etin(IDlist(5),1)/TUnit%area(IDlist(5)), TRunoff%erlateral(IDlist(5),1)/TUnit%area(IDlist(5)), TRunoff%flow(IDlist(5),1)
-    !write(unit=nio,fmt="((a10),(e20.11))") theTime, liqWater%flow(ii)
-    !write(unit=nio,fmt="((a10),6(e20.11))") theTime, liqWater%qsur(ii), liqWater%qsub(ii), liqWater%etin(ii)/(TUnit%area(ii)*TUnit%frac(ii)), liqWater%erlateral(ii)/(TUnit%area(ii)*TUnit%frac(ii)), liqWater%erin(ii), liqWater%flow(ii)
-    !if(liqWater%yr(ii) > 0._r8) then
-    !    write(unit=nio,fmt="((a10),6(e20.11))") theTime, liqWater%mr(ii)/liqWater%yr(ii),liqWater%yr(ii), liqWater%vr(ii), liqWater%erin(ii), liqWater%erout(ii)/(TUnit%area(ii)*TUnit%frac(ii)), liqWater%flow(ii)
+    !nt = 1
```
Comment (bishtgautam) [inline components/mosart/src/riverroute/MOSART_physics_mod.F90]: Please delete the commented out lines

File: components/mosart/src/riverroute/RtmHistFlds.F90 [hunk 28]
```diff
@@ -126,6 +146,61 @@ subroutine RtmHistFldsInit()
...
+         avgflag='A', long_name='WRM deficit ', &
+         ptr_rof=StorWater%deficit, default='active')
+
+      call RtmHistAddfld (fname='WRM_STORAGE', units='m3',  &
+         avgflag='A', long_name='WRM storage ', &
+         ptr_rof=StorWater%storageG, default='active')
+    endif
+!#endif
+
+!#ifdef INCLUDE_INUND
```
Comment (bishtgautam) [inline components/mosart/src/riverroute/RtmHistFlds.F90]: Please delete the commented out lines

File: components/mosart/src/riverroute/RtmHistFlds.F90 [hunk 29]
```diff
@@ -126,6 +146,61 @@ subroutine RtmHistFldsInit()
...
+    if (inundflag) then
+      call RtmHistAddfld (fname='FLOODPLAIN_VOLUME', units='m3',  &
+         avgflag='A', long_name='MOSART floodplain water volume', &
+         ptr_rof=rtmCTL%inundwf, default='active')
+      call RtmHistAddfld (fname='FLOODPLAIN_DEPTH', units='m',  &
+         avgflag='A', long_name='MOSART floodplain water depth', &
+         ptr_rof=rtmCTL%inundhf, default='active')
+        !!!!!!!!!!! added by Tian Dec 2017
```
Comment (bishtgautam) [inline components/mosart/src/riverroute/RtmHistFlds.F90]: Please delete the commented out lines

File: components/mosart/src/riverroute/RtmMod.F90 [hunk 30]
```diff
@@ -308,6 +477,18 @@ subroutine Rtmini(rtm_active,flood_active)
        RETURN
     end if
 
+!#ifdef INCLUDE_INUND
```
Comment (bishtgautam) [inline components/mosart/src/riverroute/RtmMod.F90]: Please delete commented out lines

File: components/mosart/src/riverroute/RtmMod.F90 [hunk 31]
```diff
@@ -1811,59 +2465,221 @@ subroutine Rtmrun(rstwr,nlend,rdate)
...
+       if (wrmflag) then
+          nt = 1
+          do nr = rtmCTL%begr,rtmCTL%endr
+             budget_terms(bv_dsupp_f,nt) = budget_terms(bv_dsupp_f,nt) + StorWater%supply(nr)
+             ! convert supply from m3 per coupling delta (3hrs)  to mm/s (N. Sun)
+             if (StorWater%supply(nr) > 0) then            
+               StorWater%supply(nr) = StorWater%supply(nr)/delt_coupling               
+             endif
+			 
+			  !!!!!!!!!!!!!!!!!!
```
Comment (bishtgautam) [inline components/mosart/src/riverroute/RtmMod.F90]: Please delete commented out lines

File: components/mosart/src/riverroute/RtmMod.F90 [hunk 32]
```diff
@@ -2613,30 +3970,30 @@ subroutine MOSART_init
...
-        Tunit%areatotal2(nr) = Tunit%areatotal2(nr) + avdst_eroutUp%rAttr(1,cnt)
+        Tunit%areatotal2(nr) = Tunit%areatotal2(nr) + avdst_upstrm%rAttr(1,cnt)
         areatot_tmp = areatot_tmp + Tunit%areatotal2(nr)
      enddo
      call shr_mpi_sum(areatot_tmp, areatot_new, mpicom_rof, 'areatot_new', all=.true.)
 
-     if (masterproc) then
-        write(iulog,*) trim(subname),' areatot calc ',tcnt,areatot_new
-     endif
+     !if (masterproc) then
```
Comment (bishtgautam) [inline components/mosart/src/riverroute/RtmMod.F90]: Please delete commented out lines

File: components/mosart/src/wrm/WRM_modules.F90 [hunk 33]
```diff
@@ -489,27 +583,43 @@ end subroutine WRM_storage_targets
...
 
      match = 0
+     call get_curr_date(yr, month, day, tod)
 
      damID = WRMUnit%INVicell(iunit) 
-     if ( damID > TWRMctl%NDam .OR. damID <= 0 ) then
-        print*, "Error in Regulation with DamID ",damID
-        stop
+     isDam = WRMUnit%isDam(iunit)
+!d     write(iulog,*) subname,' tcx1 ',isDam,iam,damID
```
Comment (bishtgautam) [inline components/mosart/src/wrm/WRM_modules.F90]: Please delete commented out lines

File: components/mosart/src/wrm/WRM_modules.F90 [hunk 34]
```diff
@@ -489,27 +583,43 @@ end subroutine WRM_storage_targets
...
+
+     !---------------------------------------------
+     !--- return and do nothing under certain conditions
+     !---------------------------------------------
+
+     if (damID > ctlSubwWRM%LocalNumDam .OR. damID <= 0 .or. WRMUnit%MeanMthFlow(damID,13) <= 0.01_r8) then
+        return
+        !write(iulog,*) "Error in Regulation with DamID ",damID
```
Comment (bishtgautam) [inline components/mosart/src/wrm/WRM_modules.F90]: Please delete commented out lines

File: components/mosart/src/wrm/WRM_modules.F90 [hunk 35]
```diff
@@ -559,52 +681,515 @@ subroutine Regulation(iunit, TheDeltaT)
      !  print*, "WARNING, POTENTIAL EVAP LARGER THEN RESERVOIR STORAGE, iunit ", iunit, evap
      !endif
      ! commented out because evap need to be takren into consideration in the releases
+
+
+!test NV
```
Comment (bishtgautam) [inline components/mosart/src/wrm/WRM_modules.F90]: Please delete commented out lines

File: components/mosart/src/wrm/WRM_modules.F90 [hunk 36]
```diff
@@ -559,52 +681,515 @@ subroutine Regulation(iunit, TheDeltaT)
...
+        !---------------------------
+
+        call mct_aVect_zero(aVect_wg)
+        cnt = 0
+        do iunit = begr,endr
+           cnt = cnt + 1
+           aVect_wg%rAttr(1,cnt) = StorWater%demand(iunit)
+        enddo
+
+!tcx debug
```
Comment (bishtgautam) [inline components/mosart/src/wrm/WRM_modules.F90]: Please delete commented out lines

File: components/mosart/src/wrm/WRM_modules.F90 [hunk 37]
```diff
@@ -559,52 +681,515 @@ subroutine Regulation(iunit, TheDeltaT)
...
+           call mpi_barrier(mpicom_rof,ierror)
+           call t_stopf('moswrm_ERFlow_avmult_barrier')
+        endif
+
+        call mct_aVect_zero(aVect_wd)
+        call t_startf('moswrm_ERFlow_avmult')
+        call mct_sMat_avMult(aVect_wg, sMatP_g2d, aVect_wd)
+        call t_stopf('moswrm_ERFlow_avmult')
+
+!tcx debug
```
Comment (bishtgautam) [inline components/mosart/src/wrm/WRM_modules.F90]: Please delete commented out lines

File: components/mosart/src/wrm/WRM_modules.F90 [hunk 38]
```diff
@@ -559,52 +681,515 @@ subroutine Regulation(iunit, TheDeltaT)
...
+           call t_startf('moswrm_ERFlow_bcast_barrier')
+           call mpi_barrier(mpicom_rof,ierror)
+           call t_stopf('moswrm_ERFlow_bcast_barrier')
+        endif
+
+        call t_startf('moswrm_ERFlow_bcast')
+        call mct_aVect_bcast(aVect_wdG,mastertask,mpicom_rof)
+        call t_stopf('moswrm_ERFlow_bcast')
+
+!tcx debug
```
Comment (bishtgautam) [inline components/mosart/src/wrm/WRM_modules.F90]: Please delete commented out lines

File: components/mosart/src/wrm/WRM_subw_IO_mod.F90 [hunk 39]
```diff
@@ -34,295 +54,439 @@ subroutine WRM_init
...
      endr = rtmCTL%endr
 
-     if (masterproc) write(iulog,*) subname,' begr endr = ',begr,endr
+     if (masterproc) write(iulog,FORMI) subname,' begr endr = ',iam,begr,endr
      call shr_sys_flush(iulog)
 
-     ctlSubwWRM%paraPath = '/pic/scratch/tcraig/wm_data/'
-     ctlSubwWRM%paraFile = 'US_reservoir_8th_NLDAS2.nc'
-     ctlSubwWRM%demandPath = '/pic/scratch/tcraig/wm_data/NLDAS2_GCAM_water_demand_'
+     !paraFile = '/pic/scratch/tcraig/wm_data/US_reservoir_8th_NLDAS2.nc'
```
Comment (bishtgautam) [inline components/mosart/src/wrm/WRM_subw_IO_mod.F90]: Please delete commented out lines

File: components/mosart/src/wrm/WRM_subw_IO_mod.F90 [hunk 40]
```diff
@@ -34,295 +54,439 @@ subroutine WRM_init
...
+     !-------------------
+     !--- initialize dam/gridcell interactions
+     !--- g2d will sum gridcell data to dams
+     !--- d2g will sum dam data to gridcells
+     !-------------------
 
      allocate (WRMUnit%dam_Ndepend(ctlSubwWRM%localNumDam))
-     WRMUnit%dam_Ndepend = 0        !
+     WRMUnit%dam_Ndepend = 0
+! tcraig, not on dataset anymore, compute on the fly below
```
Comment (bishtgautam) [inline components/mosart/src/wrm/WRM_subw_IO_mod.F90]: Please delete commented out lines

File: components/mosart/src/wrm/WRM_subw_IO_mod.F90 [hunk 41]
```diff
@@ -541,6 +709,11 @@ subroutine WRM_init
            if (masterproc) write(iulog,FORMR) trim(subname),' read Qmon',minval(WRMUnit%MeanMthFlow(:,mth)),maxval(WRMUnit%MeanMthFlow(:,mth))
            call shr_sys_flush(iulog)
         enddo
+!NV not okay at all but needed for checking as too much flow comng in with
```
Comment (bishtgautam) [inline components/mosart/src/wrm/WRM_subw_IO_mod.F90]: Please delete commented out lines

File: components/mosart/src/wrm/WRM_subw_IO_mod.F90 [hunk 42]
```diff
@@ -621,32 +808,123 @@ subroutine WRM_init
...
-              if ( idepend <= 0 ) then
-                 write(iulog,'(2a,4i12)') subname," Error: checking dependency, zero idepend", idam, WRMUnit%dam_Ndepend(idam), ng, idepend
-                 call shr_sys_abort(subname//' ERROR: zero idepend')
-              endif
-!tcx this should be accumlating global data, not just local data, idepend could be outside begr:endr
-!              WRMUnit%TotStorCapDepend(idepend) = WRMUnit%TotStorCapDepend(idepend) + WRMUnit%StorCap(idam)
-!              WRMUnit%TotInflowDepend(idepend) = WRMUnit%TotInflowDepend(idepend) + WRMUnit%MeanMthFlow(idam,13)
-           end do
-        end do
+!        do idam = 1,ctlSubwWRM%localNumDam
```
Comment (bishtgautam) [inline components/mosart/src/wrm/WRM_subw_IO_mod.F90]: Please delete commented out lines

File: components/mosart/bld/namelist_files/namelist_defaults_mosart.xml (hunk 1 above)
Comment (hydrotian) [reply components/mosart/bld/namelist_files/namelist_defaults_mosart.xml]: Added to Cori:/global/project/projectdirs/acme/inputdata/rof/mosart
Together with this file (for NLDAS), I also added one fore global half degree.

**Comment by rljacob** [comment]
@hydrotian you need to make the change requested by @bishtgautam above.

**Comment by bishtgautam** [comment]
SBeTR module has been reverted and this PR is ready now.

**Comment by jqyin** [comment]
Merged #3012

File: components/mosart/src/riverroute/RtmMod.F90 [hunk 43]
```diff
@@ -1679,38 +2230,41 @@ subroutine Rtmrun(rstwr,nlend,rdate)
...
+       if (inundflag) then
+          call t_startf('mosartr_inund_sim')
+!#ifdef INCLUDE_INUND
+          call MOSARTinund_simulate ( )
+!#endif     
+          call t_stopf('mosartr_inund_sim')
        else
           call t_startf('mosartr_euler')
+          ! debug 
+          write(iulog,*) 'clm-mosart subT: (call Euler) ns=', ns
```
Comment (amametjanov) [inline components/mosart/src/riverroute/RtmMod.F90]: Please comment out or surround this line with `#ifndef -DNDEBUG ... #endif` to activate it for debug runs. A 5-day run produced 122721 writes to e3sm.log.

File: components/mosart/src/riverroute/RtmTimeManager.F90 [hunk 44]
```diff
@@ -687,18 +693,47 @@ end subroutine timemgr_print
...
+           endif
+        endif
+     else
+        if (tod1 /= 0._r8) then
+           if (yy1 /= yy0) new_year = .true.
+           if (mm1 /= mm0) new_month = .true.
+           if (dd1 /= dd0) new_day = .true.
+        endif
+     endif
+     write(iulog,'(2a,4i6,3l4)') sub," yy,mm,dd = ",yy1,mm1,dd1,tod1,new_year,new_month,new_day
```
Comment (amametjanov) [inline components/mosart/src/riverroute/RtmTimeManager.F90]: Please comment out or surround this `write` with `#ifndef -DNDEBUG ... #endif`.

File: components/mosart/src/riverroute/RtmMod.F90 (hunk 43 above)
Comment (rljacob) [reply components/mosart/src/riverroute/RtmMod.F90]: @amametjanov open an issue about this.  Assign it to @hydrotian

**Pull request merged**